test/src/blockMonitor/blockchainMachine.test.ts

```diff
@@ -0,0 +1,169 @@
+        const bp: any = new EventEmitter();
+        bp.blockCache = blockCache;
+        blockProcessor = bp as BlockProcessor<IBlockStub>;
+    });
+
+    it("addComponent throws ApplicationError if already started", async () => {
+        const bm = new BlockchainMachine(blockProcessor);
+        await bm.start();
+
+        expect(() => bm.addComponent(new ExampleComponent(reducer))).to.throw(ApplicationError);
+
+        await bm.stop();
+    });
+
+    it("processNewBlock computes the initial state if the parent is not in cache", async () => {
+        const bm = new BlockchainMachine(blockProcessor);
+        bm.addComponent(new ExampleComponent(reducer));
+        await bm.start();
+
+        blockProcessor.emit(BlockProcessor.NEW_BLOCK_EVENT, blocks[0]);
+
+        verify(spiedReducer.getInitialState(anything())).once();
+        verify(spiedReducer.reduce(anything(), anything())).never();
+
+        await bm.stop();
+    });
+
+    it("processNewBlock computes state with reducer and its parent's initial state if the parent's state is not known", async () => {
+        const bm = new BlockchainMachine(blockProcessor);
+        bm.addComponent(new ExampleComponent(reducer));
+        await bm.start();
+
+        blockProcessor.emit(BlockProcessor.NEW_BLOCK_EVENT, blocks[1]);
+
+        // initializer and reducer should both be called once
+        verify(spiedReducer.getInitialState(anything())).once();
+        verify(spiedReducer.reduce(anything(), anything())).once();
+
+        // Check that the reducer was called on the right data
+        const [state, block] = capture(spiedReducer.reduce).last();
+        expect(state).to.deep.equal(initialState);
+        expect(block).to.deep.equal(blocks[1]);
+
+        await bm.stop();
+    });
+
+    it("processNewBlock computes the state with the reducer if the parent's state is known", async () => {
+        const bm = new BlockchainMachine(blockProcessor);
+        bm.addComponent(new ExampleComponent(reducer));
+        await bm.start();
+
+        blockProcessor.emit(BlockProcessor.NEW_BLOCK_EVENT, blocks[0]);
+        blockProcessor.emit(BlockProcessor.NEW_BLOCK_EVENT, blocks[1]);
+
+        resetCalls(spiedReducer);
+
+        // State of the parent is { someNumber: 42 + blocks[1].number }
+
+        blockProcessor.emit(BlockProcessor.NEW_BLOCK_EVENT, blocks[2]);
+
+        verify(spiedReducer.getInitialState(anything())).never();
+        verify(spiedReducer.reduce(anything(), anything())).once();
+
+        // Check that the reducer was called on the right data
+        const [state, block] = capture(spiedReducer.reduce).last();
+        expect(state).to.deep.equal({
+            someNumber: 42 + blocks[1].number
+        });
+        expect(block).to.deep.equal(blocks[2]);
+
+        await bm.stop();
+    });
+
+    it("processNewHead correctly calls handleChanges on the component", async () => {
```

Comment:
You can split this into multiple tests, maybe:
it("processNewHead does not call handleChanges for empty previous block")
it("processNewHead does not call handleChanges for empty previous state")
it("processNewHead does not call handleChanges for empty current state")
it("processNewHead does call handleChanges")
it("processNewHead does call handleChanges on multiple components")

Reply:
Done in bdaadf3d79c7a1b20f3f516aaec3f7db18f7a278; not exactly one-to-one on the tests above, but I think it covers all cases (except "empty previous/current state", as I added a guard in b642b9b4c1039c8da1c743f240b13e9c7f06cfea to make sure that state is never empty in processNewHead).